### PR TITLE
vendor: Add crate vendoring to the project

### PR DESCRIPTION
**Pull request by sboeuf**
By vendoring the crates, the project is not dependent on changes that
would be merged on any upstream rust-vmm crate.

Vendoring crates is important for being able to reproduce consistent
builds, which is needed by the continuous integration (CI).

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

### CODE REVIEW

**Comment by sboeuf** [comment]
@sameo @rbradford 
I think this PR can be closed, and that we actually don't need to setup any vendoring in the context of Rust. The whole point of vendoring in the context of a Go project was because there was no dependency management tool included with Go (of course there's go modules now).
In the context of Rust, `cargo` is the dependency management tool, and we should be able to lock the source version for each crate, using a git repo and a specific revision number.
The part of the vendoring that is added by `cargo vendor` is the ability to pull the sources locally, and by modifying `.cargo/config`, be able to build based on the `vendor` directory (or any other source where the vendored sources are).
So I feel the vendoring in Rust can help with local development when multiple crates modifications are involved, but otherwise `Cargo.toml` already provides the locking on specific versions.

Let me know what you think but the conclusion is that we should lock all our Cargo.toml to point to specific revisions, and that's how we will be able to achieve what we know from Go about vendoring.

**Comment by rbradford** [comment]
@sboeuf so Cargo.lock works fine until, like today, someone force pushes to a branch we depend on, causing our build to break...so maybe vendoring is a great idea for that.

**Comment by sameo** [comment]
@sboeuf And Cargo.lock is a read-only file, or at least it's supposed to be.

**Comment by sameo** [comment]
Moreover, according to:
https://doc.rust-lang.org/cargo/reference/source-replacement.html

this seems to be the right tool for what we're trying to solve...

**Comment by sboeuf** [comment]
@sameo I force pushed on top of your previous force push to add a fix to `.travis.yml`. The fix is about ignoring `vendor` directory for the `rustfmt` check.

**Comment by sboeuf** [comment]
@sameo 
```
The command "cargo clippy --all-targets --all-features -- -D warnings" exited with 101.
```
We have a problem here as clippy does not let us ignore specific directories...
```
The command "cargo build --release" exited with 101.
```
I think we need `.cargo/config` file here, so that cargo relies on the vendored files to build the project.

File: .cargo/config [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+branch = "master"
+replace-with = "vendored-sources"
+
+[source."https://github.com/rust-vmm/vm-memory"]
+git = "https://github.com/rust-vmm/vm-memory"
+branch = "master"
+replace-with = "vendored-sources"
+
+[source.vendored-sources]
+directory = "/home/samuel/devlp/hypervisor/cloud-hypervisor/vendor"
```
Comment (sameo) [inline .cargo/config]: That's a problem here...

**Comment by sameo** [comment]
Fixed now.

**Comment by sameo** [comment]
The vendoring commit is ready.
TODO:

* [ ] Document the vendoring process

**Comment by sameo** [comment]
@sboeuf @rbradford @chao-p This is ready to review.

File: docs/vendoring.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+### Locking a git dependency
+
+Some crates may depend on non published crates, that are developed and
+maintained in some git repository.
+
+Let's take the `vm-memory` crate as one example. It is a `rust-vmm` crate that
+is not yet published in `crates.io`. Several `cloud-hypervisor` crates depend
+on it. Provided that none of those dependent crates rely on a specific branch
+or revision of the `vm-memory` crate, we may want to lock our project to the
+crate revision `281b8bd6cd2927f7a65130194b203a1c2b0ad2e3`.
```
Comment (andreeaflorescu) [inline docs/vendoring.md]: How are you handling push forces with revisions?

Comment (sameo) [reply docs/vendoring.md]: Once it's vendored, a push force into the upstream repo is not going to break our build.

Comment (andreeaflorescu) [reply docs/vendoring.md]: Oh kewl, I just saw that is't basically a local cache.

**Comment by sameo** [comment]
> This looks good, only question i have is the correct terminology "locking" or "pinning" ?

Switched to "pinning"

**Pull request merged**